### PR TITLE
Modifies the rendering of constant values in rustdoc.

### PR DESCRIPTION
**Pull request by sladyn98**
Previously, non-literal expressions (like 50 + 50) would be printed alongside their value as a comment (like  = _; // 100i32). This approach often resulted in confusing documentation, especially for complex expressions.

To improve clarity, this commit changes the rendering to display the evaluated value of the expression directly (like  = 100i32;) unless the constant is a literal expression. The original expression is still included as a comment if it
doesn't match the evaluated value.

These changes aim to improve the readability and usefulness of constant value documentation generated by rustdoc.

### CODE REVIEW

**Comment by sladyn98** [comment]
CC @fmease

File: src/librustdoc/html/render/print_item.rs [hunk 1]
```diff
@@ -1491,22 +1491,23 @@ fn item_constant(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, c: &cle
...
+            write!(w, " = {value};", value = Escape(value));
         } else {
             w.write_str(";");
         }
 
-        if !is_literal {
-            if let Some(value) = &value {
-                let value_lowercase = value.to_lowercase();
-                let expr_lowercase = expr.to_lowercase();
+        let value_lowercase = value.as_ref().map(|s| s.to_lowercase());
```
Comment (fmease) [inline src/librustdoc/html/render/print_item.rs:1502]: In your earlier PR I meant to suggest removing the *entire* branch including the body (the `//` business) not just the if-condition, sorry if that wasn't clear. This here still prints `3i32 // 3i32` for `1 + 2` if I'm not mistaken.

Comment (sladyn98) [reply src/librustdoc/html/render/print_item.rs]: Correct that makes sense, sorry if I might not have understood it earlier. 

```
13: @!hasraw check failed
	`PATTERN` did not match
	// @!hasraw show_const_contents/constant.CONST_I32_HEX.html ';'
50: @hasraw check failed
	`PATTERN` did not match
	// @hasraw show_const_contents/constant.PI.html '= 3.14159265358979323846264338327950288f32;'
51: @hasraw check failed
	`PATTERN` did not match
	// @hasraw show_const_contents/constant.PI.html '; // 3.14159274f32'

```
These tests still seem to be failing, should the code be suggesting something else, I am a bit confused on what the correct comment should be here

**Comment by fmease** [comment]
Wondering if you need help with the failing test or is everything is all good?

**Comment by sladyn98** [comment]
@fmease Yeah I am not sure how to actually make it pass :  )

**Comment by fmease** [comment]
> Yeah I am not sure how to actually make it pass : )

Here's a diff:

<details>

```patch
diff --git a/tests/rustdoc/show-const-contents.rs b/tests/rustdoc/show-const-contents.rs
index 96a47218e0a..28a63de4b75 100644
--- a/tests/rustdoc/show-const-contents.rs
+++ b/tests/rustdoc/show-const-contents.rs
@@ -2,53 +2,43 @@
 // documentation.
 
 // @hasraw show_const_contents/constant.CONST_S.html 'show this'
-// @!hasraw show_const_contents/constant.CONST_S.html '; //'
 pub const CONST_S: &'static str = "show this";
 
 // @hasraw show_const_contents/constant.CONST_I32.html '= 42;'
-// @!hasraw show_const_contents/constant.CONST_I32.html '; //'
 pub const CONST_I32: i32 = 42;
 
 // @hasraw show_const_contents/constant.CONST_I32_HEX.html '= 0x42;'
-// @!hasraw show_const_contents/constant.CONST_I32_HEX.html ';'
 pub const CONST_I32_HEX: i32 = 0x42;
 
 // @hasraw show_const_contents/constant.CONST_NEG_I32.html '= -42;'
-// @!hasraw show_const_contents/constant.CONST_NEG_I32.html '; //'
 pub const CONST_NEG_I32: i32 = -42;
 
 // @hasraw show_const_contents/constant.CONST_EQ_TO_VALUE_I32.html '= 42i32;'
-// @!hasraw show_const_contents/constant.CONST_EQ_TO_VALUE_I32.html '// 42i32'
 pub const CONST_EQ_TO_VALUE_I32: i32 = 42i32;
 
 // @hasraw show_const_contents/constant.CONST_CALC_I32.html '= 43i32'
 pub const CONST_CALC_I32: i32 = 42 + 1;
 
 // @!hasraw show_const_contents/constant.CONST_REF_I32.html '= &42;'
-// @!hasraw show_const_contents/constant.CONST_REF_I32.html '; //'
 pub const CONST_REF_I32: &'static i32 = &42;
 
 // @hasraw show_const_contents/constant.CONST_I32_MAX.html '= 2_147_483_647i32'
 pub const CONST_I32_MAX: i32 = i32::MAX;
 
 // @!hasraw show_const_contents/constant.UNIT.html '= ();'
-// @!hasraw show_const_contents/constant.UNIT.html '; //'
 pub const UNIT: () = ();
 
 pub struct MyType(i32);
 
 // @!hasraw show_const_contents/constant.MY_TYPE.html '= MyType(42);'
-// @!hasraw show_const_contents/constant.MY_TYPE.html '; //'
 pub const MY_TYPE: MyType = MyType(42);
 
 pub struct MyTypeWithStr(&'static str);
 
 // @!hasraw show_const_contents/constant.MY_TYPE_WITH_STR.html '= MyTypeWithStr("show this");'
-// @!hasraw show_const_contents/constant.MY_TYPE_WITH_STR.html '; //'
 pub const MY_TYPE_WITH_STR: MyTypeWithStr = MyTypeWithStr("show this");
 
-// @hasraw show_const_contents/constant.PI.html '= 3.14159265358979323846264338327950288f32;'
-// @hasraw show_const_contents/constant.PI.html '; // 3.14159274f32'
+// @hasraw show_const_contents/constant.PI.html '= 3.14159274f32;'
 pub use std::f32::consts::PI;
 
 // @hasraw show_const_contents/constant.MAX.html '= 2_147_483_647i32'

```

</details>

As seen in the diff I've removed most `@!hasraw` checks not necessarily because they were required to make the test pass (there were only 3 test failures) but since they've become obsolete.

**Review by fmease — APPROVED** [review]
Looks good from my side. @GuillaumeGomez Do the broader changes still make sense to you (removal of `//`, showing the actual value of `PI` etc.)? TTBOMK, they are a natural consequence of fixing this FIXME. Edit: Maybe the commits should be squashed though?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I think it's fine. Just in case, does it look ok to you @jsha, @Manishearth ?

**Comment by fmease** [comment]
Manishearth's view is quite important and I agree with him to a large extend. We've had extensive discussion about this in the past. Intent (unevaluated expression) vs. representation (evaluated expression), e.g. on Zulip, topic *const value* (iirc), and on my blocked PR #99688. Initially I was a big proponent of evaluating consts but over time I've shifted my opinion slightly and I'm no longer sure what the best way forward is. That's why I haven't written any RFC yet (as promised in #99688), though I also feel a bit guilty about it.

Looking at other parts of rustdoc, we also don't normalize assoc tys for example. There's `-Znormalize-docs`, maybe we could gate this PR's behavior behind it?? As a temporary solution until we intro a new `#[doc]` attribute or similar.

Sorry sladyn98 that you stumbled into this controversial topic.

**Comment by Manishearth** [comment]
Yeah I'm extremely in favor of a new attribute here because I think it's important to give people control over this.

**Comment by sladyn98** [comment]
No worries let me know if you want to close this PR in favor of another one, or need any more help adding stuff. Thanks

**Comment by Manishearth** [comment]
I think it's probably fine to close this PR for now and instead join the Zulip discussion led by @fmease [here](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/new.20const.20display)

**Comment by sladyn98** [comment]
Closed

**Pull request closed without merging**